examples/error_estimates_forces.jl

```diff
@@ -0,0 +1,184 @@
+end
+function apply_inv_M(φk, Pk, δφnk, n)
+    δφnk = DFTK.proj_tangent_kpt(δφnk, φk)
+    function op(x)
+        apply_M(φk, Pk, x, n)
+    end
+    function f_ldiv!(x, y)
+        x .= DFTK.proj_tangent_kpt(y, φk)
+        x ./= (Pk.mean_kin[n] .+ Pk.kin)
+        DFTK.proj_tangent_kpt!(x, φk)
+    end
+    J = LinearMap{eltype(φk)}(op, size(δφnk, 1))
+    δφnk = cg(J, δφnk, Pl=DFTK.FunctionPreconditioner(f_ldiv!),
+              verbose=false, reltol=0, abstol=1e-15)
+    DFTK.proj_tangent_kpt(δφnk, φk)
+end
+function apply_metric(φ, P, δφ, A::Function)
+    map(enumerate(δφ)) do (ik, δφk)
+        Aδφk = similar(δφk)
+        φk = φ[ik]
+        for n = 1:size(δφk,2)
+            Aδφk[:,n] = A(φk, P[ik], δφk[:,n], n)
+        end
+        Aδφk
+    end
+end
+Mres = apply_metric(ψr, P, res, apply_inv_M);
+# We can now compute the modified residual ``R_{\rm Schur}(P)`` using a Schur
+# complement to approximate the error on low-frequencies[^CDKL2021]:
+# ```math
+# \begin{bmatrix} (\bm \Omega + \bm K)_{11} & (\bm \Omega + \bm K)_{12} \\ 0 & {\bm M}_{22} \end{bmatrix} \begin{bmatrix} P_{1} - P_{*1}\\ P_{2}-P_{*2} \end{bmatrix}= \begin{bmatrix} R_{1}\\R_{2} \end{bmatrix}.
+# ```
+# - Compute the projection of the residual onto the high and low frequencies:
+resLF = DFTK.transfer_blochwave(res, basis_ref, basis)
+resHF = res - DFTK.transfer_blochwave(resLF, basis, basis_ref);
+# - Compute ``{\bm M}^{-1}_{22}R_2(P)``:
+e2 = apply_metric(ψr, P, resHF, apply_inv_M);
+# - Compute the right hand side of the Schur system:
+## Rayleigh coefficients needed for `apply_Ω`
+Λ = map(enumerate(ψr)) do (ik, ψk)
+    Hk = ham.blocks[ik]
+    Hψk = Hk * ψk
+    ψk'Hψk
+end
+ΩpKe2 = DFTK.apply_Ω(e2, ψr, ham, Λ) .+ DFTK.apply_K(basis_ref, e2, ψr, ρr, occupation)
+ΩpKe2 = DFTK.transfer_blochwave(ΩpKe2, basis_ref, basis)
+rhs = resLF - ΩpKe2;
+# - Solve the Schur system to compute ``R_{\rm Schur}(P)``: this is the most costly step, but inverting ``\bm{\Omega} + \bm{K}`` on the small space has the same cost than the full SCF cycle on the small grid.
+e1 = DFTK.solve_ΩplusK(basis, ψ, rhs, occupation; tol_cg=tol)
+e1 = DFTK.transfer_blochwave(e1, basis, basis_ref)
+res_schur = e1 + Mres;
+
+# ## Error estimates
+# We start with different estimations of the forces on the first ``\rm Ti`` atom, in direction 1.
+# - Forces computed with the reference solution:
+f_ref = compute_forces(scfres_ref)
+println("F(P_*) = $(f_ref[1][1][1])")
+# - Forces computed with the variational approximation:
+f = compute_forces(scfres)
+println("F(P) = $(f[1][1][1])")
+# We then try to improve ``F(P)`` using the first order linearization:
+# ```math
+# F(P) = F(P_*) + {\rm d}F(P)\cdot(P-P_*).
+# ```
+# To this end, we use the `ForwardDiff.jl` package to compute ``{\rm d}F(P)``
+# using automatic differentiation.
+function df(basis, occupation, ψ, δψ)
+    δρ = DFTK.compute_δρ(basis, ψ, δψ, occupation)
+    function f(ε)
+        compute_forces(basis, ψ .+ ε.*δψ, occupation; ρ=ε.*δρ)
```

Comment:
I'm confused by the rho here. Can't you just do rho = compute_rho(basis, psi+eps dpsi)?

Reply:
Hum, I'm not convinced. With that there is no need to recompute rho every time f is being called, why do you find it confusing ?

Reply:
OK on reflection I agree it's always true for any reasonable case. Still, computing drho by compute_drho is basically doing the job of the AD, no reason to do it

Reply:
So, when I do that with
```
function df(basis, occupation, ψ, δψ)                                                                            
    function f(ε)                                                               
        compute_forces(basis, ψ .+ ε.*δψ, occupation;  ρ=compute_density(basis, ψ .+ ε.*δψ, occupation))        
    end                                                                         
    ForwardDiff.derivative(f, 0)                                                
end; 
```
then I have again this type error for ForwardDiff mentioned above.

Reply:
Hm interesting. Because T now has to be eltype(rho)? So we have to promote T from rho, psi and basis?!

Reply:
Well, I don't know... In https://github.com/JuliaMolSim/DFTK.jl/blob/master/examples/forwarddiff.jl, it's `compute_δρ` which is being used.

Reply:
I see. I thought we had a ForwardDiff frule for compute_rho anyway, but apparently we don't. In this case as a workaround you can just do what you do now but with  ρ=ρ+ε.*δρ

Reply:
> I see. I thought we had a ForwardDiff frule for compute_rho anyway, but apparently we don't. 
not yet

> as a workaround you can just do what you do now

can you also add a comment for that. We should really have the frule.

Reply:
or just make compute_rho compatible with forwarddiff anyway?

Reply:
Don't we have the symmetry issue to work around in that case. I don't remember or was this reverse diff?

Reply:
reverse diff is definitely tricky because we mutate a lot, but forward diff should be OK